cmds/id/id.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Synopsis:
+//      id [-gGnu]
+//
+// Description:
+//      id displays the uid, guid and groups of the calling process
+// Options:
+//  		-g, --group     print only the effective group ID
+//		  -G, --groups    print all group IDs
+//		  -n, --name      print a name instead of a number, for -ugG
+//		  -u, --user      print only the effective user ID
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+var (
+	GroupFile  = "/etc/group"
+	PasswdFile = "/etc/passwd"
+
+	flags struct {
+		group  bool
+		groups bool
+		name   bool
+		user   bool
+	}
+)
+
+func correctFlags(flags ...bool) bool {
+	n := 0
+	for _, v := range flags {
+		if v {
+			n += 1
+		}
+	}
+	if n > 1 {
+		return false
+	} else {
+		return true
+	}
+}
+
+func initFlags() error {
+	flag.BoolVar(&flags.group, "g", false, "print only the effective group ID")
+	flag.BoolVar(&flags.groups, "G", false, "print all group IDs")
+	flag.BoolVar(&flags.name, "n", false, "print a name instead of a number, for -ugG")
+	flag.BoolVar(&flags.user, "u", false, "print only the effective user ID")
+	flag.Parse()
+	if !correctFlags(flags.groups, flags.group, flags.user) {
+		return fmt.Errorf("cannot print \"only\" of more than one choice\n")
+
+	}
+	if flags.name && !(flags.groups || flags.group || flags.user) {
+		return fmt.Errorf("cannot print only names in default format\n")
+	}
+
+	return nil
+}
+
+type User struct {
+	name   string
+	uid    int
+	euid   int
+	gid    int
+	groups map[int]string
+}
+
+// readGroups reads the GroupFile for groups.
+// It assumes the format "name:passwd:number:groupList".
+func readGroups() (map[int]string, error) {
+	groupFile, err := os.Open(GroupFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var groupInfo []string
+
+	groupsMap := make(map[int]string)
+	groupScanner := bufio.NewScanner(groupFile)
+
+	for groupScanner.Scan() {
+		groupInfo = strings.Split(groupScanner.Text(), ":")
+		groupNum, err := strconv.Atoi(groupInfo[2])
+		if err != nil {
+			return nil, err
+		}
+
+		groupsMap[groupNum] = groupInfo[0]
+	}
+
+	return groupsMap, nil
+}
+
+func (u *User) UID() int {
+	return u.uid
+}
+
+func (u *User) GetEuid() int {
+	return u.euid
+}
+
+func (u *User) GID() int {
+	return u.gid
+}
+
+func (u *User) Name() string {
+	return u.name
+}
+
+func (u *User) Groups() map[int]string {
+	return u.groups
+}
+
+func (u *User) GIDName() string {
+	val := u.Groups()[u.UID()]
+	return val
+}
+
+// NewUser is a factory method for the User type.
+func NewUser() (*User, error) {
+	emptyMap := make(map[int]string)
+	u := &User{"", syscall.Getuid(), syscall.Geteuid(), syscall.Getgid(), emptyMap}
```

Comment:
or even for an empty map
map[int]string{}
https://play.golang.org/p/51q8-qBjzQ